File: Week 13/1.QueueImplementation.java
/*
 * WEEK 13 - QUEUES
 * Topic: Queue Implementations + Sliding Window Maximum
 *
 * QUEUE: FIFO (First In, First Out) data structure.
 * Think of a line at a store — first to arrive, first to be served.
 *
 * CORE OPERATIONS (all O(1)):
 * - enqueue(x) / offer(x): add to rear
 * - dequeue() / poll():    remove from front
 * - peek() / front():      view front element
 * - isEmpty()
 *
 * VARIANTS:
 * - Circular Queue: efficient array-based queue
 * - Deque (Double-Ended Queue): insert/delete at BOTH ends
 * - Priority Queue: element with highest priority dequeued first
 *
 * In Java: use ArrayDeque as the Queue implementation.
 *
 * APPLICATIONS:
 * 1. BFS (Week 17)
 * 2. Sliding window maximum (monotonic deque)
 * 3. Level-order tree traversal
 * 4. Task scheduling
 *
 * SLIDING WINDOW MAXIMUM (hard problem):
 * Given array of n integers and window size k,
 * find the maximum in each window of size k.
 * Naive: O(n*k). Optimal with monotonic deque: O(n).
 */

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

public class QueueImplementation {

    // Circular Queue implementation (array-based)
    static class CircularQueue {
        int[] data;
        int front, rear, size, capacity;

        CircularQueue(int capacity) {
            this.capacity = capacity;
            data = new int[capacity];
            front = rear = size = 0;
        }

        boolean enqueue(int x) {
            if (size == capacity) return false; // full
            data[rear] = x;
            rear = (rear + 1) % capacity; // wrap around
            size++;
            return true;
        }

        int dequeue() {
            if (size == 0) throw new RuntimeException("Queue empty");
            int val = data[front];
            front = (front + 1) % capacity;
            size--;
            return val;
        }

        int peek() {
            if (size == 0) throw new RuntimeException("Queue empty");
            return data[front];
        }

        boolean isEmpty() { return size == 0; }
        boolean isFull() { return size == capacity; }
    }

    // Queue using two stacks (classic interview problem)
    // amortized O(1) per operation
    static class QueueUsingStacks {
        Deque<Integer> inbox = new ArrayDeque<>();  // for enqueue
        Deque<Integer> outbox = new ArrayDeque<>(); // for dequeue

        void enqueue(int x) { inbox.push(x); }

        int dequeue() {
            if (outbox.isEmpty()) {
                // Transfer all from inbox to outbox (reverses order)
                while (!inbox.isEmpty()) outbox.push(inbox.pop());
            }
            if (outbox.isEmpty()) throw new RuntimeException("Queue empty");
            return outbox.pop();
        }

        int peek() {
            if (outbox.isEmpty()) while (!inbox.isEmpty()) outbox.push(inbox.pop());
            return outbox.peek();
        }

        boolean isEmpty() { return inbox.isEmpty() && outbox.isEmpty(); }
    }

    // Sliding Window Maximum using Monotonic Deque
    // Time: O(n), Space: O(k)
    // Maintain a deque of INDICES where values are in DECREASING order.
    // Front of deque always has the index of the maximum for current window.
    static int[] slidingWindowMax(int[] arr, int k) {
        int n = arr.length;
        if (n == 0 || k == 0) return new int[0];
        int[] result = new int[n - k + 1];
        Deque<Integer> deque = new ArrayDeque<>(); // stores indices

        for (int i = 0; i < n; i++) {
            // Remove indices outside window
            while (!deque.isEmpty() && deque.peekFirst() < i - k + 1)
                deque.pollFirst();

            // Remove indices of elements smaller than current (they can't be future maxima)
            while (!deque.isEmpty() && arr[deque.peekLast()] < arr[i])
                deque.pollLast();

            deque.offerLast(i);

            // Window is full starting at index k-1
            if (i >= k - 1) result[i - k + 1] = arr[deque.peekFirst()];
        }
        return result;
    }

    public static void main(String[] args) {
        // Circular Queue
        System.out.println("=== Circular Queue ===");
        CircularQueue cq = new CircularQueue(4);
        cq.enqueue(1); cq.enqueue(2); cq.enqueue(3); cq.enqueue(4);
        System.out.println("Full: " + cq.isFull());
        System.out.println("Dequeue: " + cq.dequeue()); // 1
        cq.enqueue(5);
        System.out.println("Peek: " + cq.peek()); // 2

        // Queue using two stacks
        System.out.println("\n=== Queue using Two Stacks ===");
        QueueUsingStacks q = new QueueUsingStacks();
        q.enqueue(1); q.enqueue(2); q.enqueue(3);
        System.out.println("dequeue: " + q.dequeue()); // 1 (FIFO)
        System.out.println("dequeue: " + q.dequeue()); // 2
        q.enqueue(4);
        System.out.println("peek: " + q.peek());       // 3
        System.out.println("dequeue: " + q.dequeue()); // 3
        System.out.println("dequeue: " + q.dequeue()); // 4

        // Java's built-in Queue
        System.out.println("\n=== Java ArrayDeque as Queue ===");
        Queue<Integer> jQueue = new LinkedList<>();
        jQueue.offer(10); jQueue.offer(20); jQueue.offer(30);
        System.out.println("peek: " + jQueue.peek());  // 10
        System.out.println("poll: " + jQueue.poll());  // 10
        System.out.println("poll: " + jQueue.poll());  // 20

        // Sliding Window Maximum
        System.out.println("\n=== Sliding Window Maximum ===");
        int[] arr = {1, 3, -1, -3, 5, 3, 6, 7};
        int k = 3;
        System.out.println("Array: " + Arrays.toString(arr));
        System.out.println("Window size k=" + k);
        System.out.println("Max in each window: " + Arrays.toString(slidingWindowMax(arr, k)));
        // Expected: [3, 3, 5, 5, 6, 7]

        int[] arr2 = {9, 11};
        System.out.println("\nArray: " + Arrays.toString(arr2) + ", k=2");
        System.out.println("Max: " + Arrays.toString(slidingWindowMax(arr2, 2)));
        // Expected: [11]
    }
}

File: Week 13/python/1.QueueImplementation.py
"""
WEEK 13 - PYTHON DSA
Topic: Queue Implementations + Sliding Window Maximum
File: 1.QueueImplementation.py

CONCEPT:
A queue is FIFO: enqueue at the back, dequeue from the front. Variants:
- Circular queue (fixed-size array with modular arithmetic).
- Queue from two stacks (classic interview problem).
- Deque (double-ended) — used here for sliding window maximum.

KEY POINTS:
- Standard ops: enqueue/offer (add rear), dequeue/poll (remove front),
  peek/front (read front), is_empty.
- For the sliding window maximum we use a monotonic deque storing INDICES
  whose values are strictly decreasing from front to back.

ALGORITHM / APPROACH:
- CircularQueue: front, rear, size; rear = (rear+1) % capacity on enqueue;
  front advances similarly on dequeue.
- QueueUsingStacks: push to inbox; on dequeue, if outbox empty, drain inbox
  into outbox (reversing order). Amortized O(1).
- slidingWindowMax: for each i:
    * pop indices from the front that fall out of the window
    * pop indices from the back whose values < arr[i] (they can never be
      the max for any current/future window)
    * push i; record arr[front] when i >= k-1.

PYTHON-SPECIFIC NOTES:
- collections.deque is a doubly-linked-list-of-blocks giving O(1) on both
  ends. It's the canonical queue/deque in Python.
- queue.Queue is for thread-safe producer/consumer use, NOT general use.
- Type hints + Optional.

DRY RUN:
Example 1: CircularQueue capacity=4
  enqueue 1,2,3,4 -> data=[1,2,3,4] front=0 rear=0 (wrap) size=4 full
  dequeue -> 1; front=1; size=3
  enqueue 5 -> data[rear=0]=5; rear=1; size=4
  peek -> data[front=1] = 2

Example 2: slidingWindowMax([1,3,-1,-3,5,3,6,7], k=3)
  i=0: deque=[0]
  i=1: pop 0 (1<3); deque=[1]; window not full
  i=2: deque=[1,2]; res[0]=arr[1]=3
  i=3: deque=[1,2,3]; res[1]=arr[1]=3
  i=4: pop 3 (-3<5); pop 2 (-1<5); pop 1 (3<5); deque=[4]; res[2]=5
  ...
  Final: [3,3,5,5,6,7]

COMPLEXITY:
  CircularQueue: O(1) per op
  QueueUsingStacks: amortized O(1) per op
  slidingWindowMax: O(n) total (each element pushed/popped at most once)
"""

from __future__ import annotations
from collections import deque
from typing import Deque, List, Optional


class CircularQueue:
    """Fixed-capacity ring buffer."""

    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.data: List[Optional[int]] = [None] * capacity
        self.front = 0
        self.rear = 0
        self.size = 0

    def enqueue(self, x: int) -> bool:
        if self.size == self.capacity:
            return False  # full
        self.data[self.rear] = x
        self.rear = (self.rear + 1) % self.capacity
        self.size += 1
        return True

    def dequeue(self) -> int:
        if self.size == 0:
            raise IndexError("Queue empty")
        v = self.data[self.front]
        assert v is not None
        self.data[self.front] = None
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return v

    def peek(self) -> int:
        if self.size == 0:
            raise IndexError("Queue empty")
        v = self.data[self.front]
        assert v is not None
        return v

    def is_empty(self) -> bool: return self.size == 0
    def is_full(self) -> bool: return self.size == self.capacity


class QueueUsingStacks:
    """Two-stacks queue. Amortized O(1) per op."""

    def __init__(self) -> None:
        self.inbox: List[int] = []
        self.outbox: List[int] = []

    def enqueue(self, x: int) -> None:
        self.inbox.append(x)

    def _shift(self) -> None:
        if not self.outbox:
            while self.inbox:
                self.outbox.append(self.inbox.pop())

    def dequeue(self) -> int:
        self._shift()
        if not self.outbox:
            raise IndexError("Queue empty")
        return self.outbox.pop()

    def peek(self) -> int:
        self._shift()
        if not self.outbox:
            raise IndexError("Queue empty")
        return self.outbox[-1]

    def is_empty(self) -> bool: return not self.inbox and not self.outbox


def sliding_window_max(arr: List[int], k: int) -> List[int]:
    """Monotonic deque solution. O(n) time, O(k) space."""
    n = len(arr)
    if n == 0 or k == 0:
        return []
    result: List[int] = []
    dq: Deque[int] = deque()  # indices, arr[dq] strictly decreasing
    for i in range(n):
        # remove indices outside window
        while dq and dq[0] < i - k + 1:
            dq.popleft()
        # remove smaller values from back
        while dq and arr[dq[-1]] < arr[i]:
            dq.pop()
        dq.append(i)
        if i >= k - 1:
            result.append(arr[dq[0]])
    return result


def main() -> None:
    print("=== Circular Queue ===")
    cq = CircularQueue(4)
    for v in [1, 2, 3, 4]: cq.enqueue(v)
    print("Full:", cq.is_full())
    print("Dequeue:", cq.dequeue())
    cq.enqueue(5)
    print("Peek:", cq.peek())

    print("\n=== Queue using Two Stacks ===")
    q = QueueUsingStacks()
    for v in [1, 2, 3]: q.enqueue(v)
    print("dequeue:", q.dequeue())
    print("dequeue:", q.dequeue())
    q.enqueue(4)
    print("peek:", q.peek())
    print("dequeue:", q.dequeue())
    print("dequeue:", q.dequeue())

    print("\n=== Python deque as Queue ===")
    dq: Deque[int] = deque()
    for v in [10, 20, 30]: dq.append(v)
    print("peek:", dq[0])
    print("poll:", dq.popleft())
    print("poll:", dq.popleft())

    print("\n=== Sliding Window Maximum ===")
    arr = [1, 3, -1, -3, 5, 3, 6, 7]
    k = 3
    print("Array:", arr, "k =", k)
    print("Max in each window:", sliding_window_max(arr, k))  # [3,3,5,5,6,7]

    arr2 = [9, 11]
    print("\nArray:", arr2, "k=2")
    print("Max:", sliding_window_max(arr2, 2))


if __name__ == "__main__":
    main()


"""
NOTES (vs. Java):
- Java's recommended Queue is ArrayDeque. The Python equivalent is
  collections.deque. Both are O(1) at both ends.
- queue.Queue is a thread-safe blocking queue (vs. java.util.concurrent
  BlockingQueue) — different beast.
- Python lists used as stacks are O(1) for append/pop at the END only;
  popping from the front is O(n). Use deque for double-ended ops.
"""
